utils/ptrace.c

```diff
@@ -0,0 +1,129 @@
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <sys/ptrace.h>
+#include <sys/user.h>
+#include <wait.h>
+#include <time.h>
+
+#include "ptrace.h"
+#include "utils.h"
+
+void ptrace_attach(pid_t target)
+{
+	int waitpidstatus;
+
+	if (ptrace(PTRACE_ATTACH, target, NULL, NULL) == -1)
+		pr_err("ptrace(PTRACE_ATTACH) failed");
+
+	if (waitpid(target, &waitpidstatus, WUNTRACED) != target)
+		pr_err("waitpid(%d) failed", target);
+}
+
+void ptrace_detach(pid_t target)
+{
+	if (ptrace(PTRACE_DETACH, target, NULL, NULL) == -1)
+		pr_err("ptrace(PTRACE_DETACH) failed");
+}
+
+void ptrace_getregs(pid_t target, struct REG_TYPE *regs)
+{
+	if (ptrace(PTRACE_GETREGS, target, NULL, regs) == -1)
+		pr_err("ptrace(PTRACE_GETREGS) failed");
+}
+
+void ptrace_cont(pid_t target)
+{
+	struct timespec *sleeptime = malloc(sizeof(struct timespec));
+
+	sleeptime->tv_sec = 1;
+	sleeptime->tv_nsec = 0000000;
+
+	if (ptrace(PTRACE_CONT, target, NULL, NULL) == -1)
+		pr_err("ptrace(PTRACE_CONT) failed");
+	nanosleep(sleeptime, NULL);
+	// make sure the target process received SIGTRAP after stopping.
+	checktargetsig(target);
+}
+
+void ptrace_setregs(pid_t target, struct REG_TYPE *regs)
+{
+	if (ptrace(PTRACE_SETREGS, target, NULL, regs) == -1)
+		pr_err("ptrace(PTRACE_SETREGS) failed");
+}
+
+siginfo_t ptrace_getsiginfo(pid_t target)
```

Comment:
Hmm.. It'd be nice if you can pass it as an argument IMO.

Reply:
@namhyung  what is mean IMO?

Reply:
It's an acronym for 'In My Opinion'..